keps/sig-network/20201014-NamespaceAsName.md

```diff
@@ -0,0 +1,341 @@
+```
+
+The latter comment received 9 likes as a github issue - indiciating the general popularity of this as a request, and that correlates well to feedback we've seen in the broader community as well.
+
+We thus conclude that, even though targetting an object using its name is not normal in K8s, in the case of NetworkPolicies, the overwhelming need for universal, easy to define security boundaries, makes a strong case for amending the API to support a "special" selector for namespace names that is **independent** of labels.
+
+### Goals
+
+- Add a `matchName` option (which is additive to the current `matchLabels` selector).
+- Add an additional semantic mechanism to "allow all" namespaces while excluding a finite, specific, list of namespaces (canonically, this might be `kube-system`, as this is an obvious security boundary which most clusters would prefer).
+
+### Non-Goals
+
+- Support matching of wildcard namespaces
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyPort`` object struct specify a new ``PortRange`` field:
+
+```
+    // NetworkPolicyPort describes a port to allow traffic on
+    type NetworkPolicyPort struct {
+        PortRange *string
+    }
+```
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an administrator I want to block all users from accessing the kube-system namespace as a fundamental default policy.
+
+#### Story 2
+
+As an user I want to "just add" a namespace to my allow list without having to manage the labels which might get added/removed over time from said namespace.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- CNIs may not choose, initially to support this construct, and dilligent communication with CNI providers will be needed to make sure its widely adopted.
+- Fractionating the Kubernetes API idiom by introducing a separate way to select objects.  We accept this cost because security is a fundamentally important paradigm that justifies breaking other paradigms, at times. 
+
+## Design Details
+
+- Add a new selector to the network policy peer data structure which can switch between allowing a "matchName" or "matchLabels" implementation, supporting a policy that is expressed like this:
+
+This selector has two, EXCLUSIVE, possible input types:
+
+- A list of conventional namespaces OR 
+- A `*` with an `exclude` rule as in input to this selector
+
+The "conventional namespaces" will look like so:
+```
+kind: NetworkPolicy
+apiVersion: networking.k8s.io/v1
+metadata:
+  name: mysql-allow-app
+spec:
+  podSelector:
+    matchLabels:
+      app: mysql
+  ingress:
+  - from:
+    - namespaceSelector:
+         matchName:
```

Comment:
Adding a `matchName` field to `namespaceSelector` is potentially a breaking change because we would have to create a new internal struct for `namespaceSelector` to something like:
```go
type NamespaceSelector struct {
  metav1.LabelSelector

  matchNames []string
}
``` 

While maybe we can get away with not breaking json/yaml serialization, I think we still break clients creating NetworkPolicy types off of the new version? If this is a breaking change, at worse we should add a new field outside namespaceSelector. But breaking changes aside, I like `matchNames` under `namespaceSelector` much better.

Reply:
An alternative to changing the syntax would be to add a "virtual label" with some name that is not a valid label name (eg, `virtual/name` which is not valid because the prefix part isn't a DNS name). Then you could just do:

    - namespaceSelector:
        matchLabels:
          virtual/name: my-frontend

but also you could do

    - namespaceSelector:
        matchExpressions:
          - key: virtual/name
            operator: NotIn
            values:
              - kube-system

Reply:
ive never understood the virtual labels thing... are there any examples of how people hack stuff up to use these ?

Reply:
Let's not invent ad hoc syntax. Please :)

Adding a field to metav1.LabelSelector is dangerous at best.  I find it confusing - it's not a selector in the usual kubernetes sense.  Also breaks clients which should not be taken lightly.

Virtual labels are interesting but if we want to go that route, we'd need to pause and define that whole mechanism.  Does it apply to all selectors and all resources?  Is it just for "name" or does it have other uses?  Does it show up when I GET an object?  This is a KEP of its own.

The *simplest* option here is to add `namespaceNames []string` to `NetworkPolicyPeer`.

You still have to define what consumers should do if they don't support this field.  They would see a `NetworkPolicyPeer` with no fields set which fails validation.  So to make this work we would need to specifically loosen that validation.  Define that to mean "allow none".  That, of course has to be alpha-gated.  We'd really need to work with the mail implementations to verify that they either do the right thing already or change them to treat that safely.

So (for example) the very best case: 1.20 could allow empty peers and add support for `namespaceNames[]` if and only if the  gate was enabled.  1.21 could flip the gate to default-on for both (beta or GA, TBD).  If it takes a while to get implementations doing the right thing, we may have to leave it in alpha for longer.

Reply:
> You still have to define what consumers should do if they don't support this field.

> If it takes a while to get implementations doing the right thing, we may have to leave it in alpha for longer.

The feature gate should be tracking "are we satisfied with the API?". That's separate from "does everyone implement the API?" though clearly related.

#2137 is an attempt to deal with the problem of NetworkPolicy feature versioning more generically (though of course, if it merges in the same release as this feature does then it doesn't really solve the problem for _this_ feature since plugins that didn't implement namespace names probably wouldn't implementing versioning either)